File: solutions/go/book-store/1/book_store.go
package bookstore

import (
	"slices"
)

func Cost(books []int) int {
	var frequencies [5]int
	for _, book := range books {
		frequencies[book - 1]++
	}

	slices.Sort(frequencies[:])

	basket := [5]int{frequencies[0]}
	for i := 1; i < 5; i++ {
		basket[i] = frequencies[i] - frequencies[i - 1]
	}

	count := min(basket[0], basket[2])
	if count > 0 {
		basket[1] += count << 1
		basket[0] -= count 
		basket[2] -= count 
	}

	return basket[0] * 3000 + basket[1] * 2560 + basket[2] * 2160 + basket[3] * 1520  + basket[4] * 800
}

File: solutions/java/book-store/2/src/main/java/BookStore.java
import java.util.Arrays;
import java.util.List;

public class BookStore {

    private static final double BASE_PRICE = 8.0;

    public double calculateBasketCost(List<Integer> books) {
        if(books.isEmpty()) return .0;

        int[] frequencies = new int[5];
        for (var book : books) {
            frequencies[book - 1]++;
        }
        Arrays.sort(frequencies);

        int[] basket = new int[5];
        basket[0] = frequencies[0];

        for(int i = 1; i < 5; i++) {
            basket[i] = frequencies[i] - frequencies[i - 1];
        }

        int min = Math.min(basket[0], basket[2]);
        if (min > 0) {
            basket[1] += min * 2;
            basket[0] -= min;
            basket[2] -= min;
        }
    
        return BASE_PRICE * (basket[0] * 5 * 0.75
                             + basket[1] * 4 * 0.8
                             + basket[2] * 3 * 0.9
                             + basket[3] * 2 * 0.95
                             + basket[4]);
    }

}
